### PR TITLE
Removing node field from machine status

### LINKED ISSUE #389: Remove MCM's dependency on Status.Node while adoption
## Improvement
Currently MCM stores node-object name at both `labels` and `Status.Node` on machine-object. We can remove the dependency from the `Status.Node`, and use only labels. This could be helpful during shoot-control plane migration, as we could simply remove the `Machine.Status` and let it be reconciled by MCM after control-plane migration. 

cc @prashanth26 @plkokanov

### PR DESCRIPTION
**Pull request by rishabh-11**
**What this PR does / why we need it**:
This PR removes the `node` field in the `machine status` and replaces its usage with the `node label` present in the machine object's metadata.

**Which issue(s) this PR fixes**:
Fixes #389 

**Special notes for your reviewer**:
ITs were run for `aws` and `azure` providers with the seed as the control cluster.

**Release note**:

```breaking user
`node` field is removed from machine status. controller will now depend on the node label which already was present in the machine object's metadata. If you(or your controller) are dependent on the `status.node` field of the machine object, then kindly use `node` label under `.metadata.labels`
```

### CODE REVIEW

File: pkg/util/provider/machinecontroller/machine.go [hunk 1]
```diff
@@ -161,7 +161,7 @@ func (c *controller) reconcileClusterMachine(ctx context.Context, machine *v1alp
 		return retry, err
 	}
 
-	if machine.Status.Node != "" {
+	if machine.Labels[v1alpha1.NodeLabelKey] != "" {
```
Comment (himanshu-kun) [inline pkg/util/provider/machinecontroller/machine.go:164]:
```suggestion
	if machine.Labels!=nil && machine.Labels[v1alpha1.NodeLabelKey] != "" && machine.Status.CurrentStatus.Phase != "" {
```

This is done to make sure that we are entering this only when the controller has updated the `phase` of the machine which is part of status. if we don't do this then we could enter this block when just the spec is updated and not status.

File: pkg/util/provider/machinecontroller/machine.go [hunk 2]
```diff
@@ -173,7 +173,7 @@ func (c *controller) reconcileClusterMachine(ctx context.Context, machine *v1alp
 			return retry, err
 		}
 	}
-	if machine.Spec.ProviderID == "" || machine.Status.CurrentStatus.Phase == "" || machine.Status.Node == "" {
+	if machine.Spec.ProviderID == "" || machine.Status.CurrentStatus.Phase == "" || machine.Labels[v1alpha1.NodeLabelKey] == "" {
```
Comment (himanshu-kun) [inline pkg/util/provider/machinecontroller/machine.go:176]:
```suggestion
	if machine.Spec.ProviderID == "" || machine.Status.CurrentStatus.Phase == "" {
```
if we go by the above suggestion of mine , then we know that if the `machine.Status.CurrentStatus.Phase` value is set then we don't need to check for the node label.

File: pkg/util/provider/machinecontroller/machine_test.go [hunk 3]
```diff
@@ -123,82 +123,6 @@ var _ = Describe("machine", func() {
...
-			Describe("Update conditions of a non-existing machine", func() {
-				It("should return error", func() {
-					stop := make(chan struct{})
-					defer close(stop)
-
-					objects := []runtime.Object{}
-					c, trackers := createController(stop, testNamespace, objects, nil, nil)
-					defer trackers.Stop()
-
-					testMachine := &v1alpha1.Machine{
```
Comment (himanshu-kun) [inline pkg/util/provider/machinecontroller/machine_test.go:137]: I think we shouldn't remove these commented codes until we are sure we have addressed whatever they suggest.

**Review by himanshu-kun — CHANGES_REQUESTED** [review]
Thanks for the changes. Sorry the ultra-delay in reviewing this PR.
I have asked for a few changes.

File: pkg/controller/machine.go [hunk 4]
```diff
@@ -314,7 +313,10 @@ func (c *controller) getMachineFromNode(nodeName string) (*v1alpha1.Machine, err
 }
 
 func (c *controller) updateMachineState(ctx context.Context, machine *v1alpha1.Machine) (*v1alpha1.Machine, error) {
-	nodeName := machine.Status.Node
+	nodeName := ""
+	if machine.Labels != nil {
```
[the remark this answers is not in the document]
Comment (unmarshall) [reply pkg/controller/machine.go:317]: actually you do not need a nil check.
```golang
func main() {
	var m map[string]string
	m = nil
	fmt.Printf("map: %v, value: %v", m == nil, m["bingo"])
}
```

[the remark this answers is not in the document]
Comment (himanshu-kun) [reply pkg/controller/machine.go:317]: then pls remove the `machine.Labels != nil` check also @rishabh-11

File: pkg/util/provider/machinecontroller/machine_test.go (hunk 3 above)
Comment (rishabh-11) [reply pkg/util/provider/machinecontroller/machine_test.go]: ok

File: pkg/controller/machine.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (rishabh-11) [reply pkg/controller/machine.go]: Done.

File: pkg/controller/machine.go [hunk 5]
```diff
@@ -333,10 +335,13 @@ func (c *controller) updateMachineState(ctx context.Context, machine *v1alpha1.M
...
-				clone.Status.Node = nodeName
-				clone, err = c.controlMachineClient.Machines(clone.Namespace).UpdateStatus(ctx, clone, metav1.UpdateOptions{})
+				if clone.Labels == nil {
+					clone.Labels = make(map[string]string)
+				}
+				clone.Labels[v1alpha1.NodeLabelKey] = nodeName
+				machine, err = c.controlMachineClient.Machines(clone.Namespace).Update(ctx, clone, metav1.UpdateOptions{})
 				if err != nil {
-					klog.Errorf("Could not update status of the machine-object %s due to error %v", machine.Name, err)
+					klog.Errorf("Could not update the machine-object %s due to error %v", machine.Name, err)
```
[the remark this answers is not in the document]
Comment (rishabh-11) [reply pkg/controller/machine.go]: done

File: pkg/controller/machine.go [hunk 6]
```diff
@@ -333,10 +335,13 @@ func (c *controller) updateMachineState(ctx context.Context, machine *v1alpha1.M
 				klog.V(2).Infof("Adopting the node object %s for machine %s", node.Name, machine.Name)
 				nodeName = node.Name
 				clone := machine.DeepCopy()
-				clone.Status.Node = nodeName
-				clone, err = c.controlMachineClient.Machines(clone.Namespace).UpdateStatus(ctx, clone, metav1.UpdateOptions{})
+				if clone.Labels == nil {
+					clone.Labels = make(map[string]string)
+				}
+				clone.Labels[v1alpha1.NodeLabelKey] = nodeName
+				machine, err = c.controlMachineClient.Machines(clone.Namespace).Update(ctx, clone, metav1.UpdateOptions{})
```
[the remark this answers is not in the document]
Comment (rishabh-11) [reply pkg/controller/machine.go]: yeah, I don't remember right now. Anyways, I have reverted back to `clone`.

File: pkg/util/provider/machinecontroller/machine.go (hunk 2 above)
Comment (rishabh-11) [reply pkg/util/provider/machinecontroller/machine.go]: got it. done

**Review by himanshu-kun — APPROVED** [review]
/lgtm

**Pull request merged**